File: app/src/main/java/com/cardee/owner_car_details/presenter/StrategyRentalDetailPresenter.java
package com.cardee.owner_car_details.presenter;

import com.cardee.data_source.Error;
import com.cardee.data_source.OwnerCarRepository;
import com.cardee.data_source.OwnerCarsRepository;
import com.cardee.domain.UseCase;
import com.cardee.domain.UseCaseExecutor;
import com.cardee.domain.owner.entity.RentalDetails;
import com.cardee.domain.owner.usecase.SaveDailyRental;
import com.cardee.domain.owner.usecase.SaveHourlyTiming;
import com.cardee.domain.owner.usecase.UpdateDailyAcceptCashState;
import com.cardee.domain.owner.usecase.UpdateDailyCurbsideDeliveryState;
import com.cardee.domain.owner.usecase.UpdateDailyInstantBookingState;
import com.cardee.domain.owner.usecase.UpdateDailyinstantBookingCount;
import com.cardee.domain.owner.usecase.UpdateHourlyAcceptCashState;
import com.cardee.domain.owner.usecase.UpdateHourlyCurbsideDeliveryState;
import com.cardee.domain.owner.usecase.UpdateHourlyInstantBookingCount;
import com.cardee.domain.owner.usecase.UpdateHourlyInstantBookingState;
import com.cardee.owner_car_details.RentalDetailsContract;

public class StrategyRentalDetailPresenter implements RentalDetailsContract.Presenter {

    private static final String TAG = StrategyRentalDetailPresenter.class.getSimpleName();

    public enum Strategy {
        DAILY, HOURLY
    }

    private final Strategy strategy;
    private final UseCaseExecutor executor;
    private final SaveHourlyTiming saveHourlyDates;
    private final SaveDailyRental saveDailyRental;
    private RentalDetailsContract.ControlView view;
    private RentalDetails rentalDetails;

    public StrategyRentalDetailPresenter(RentalDetailsContract.ControlView view, Strategy strategy) {
        this.strategy = strategy;
        this.view = view;
        executor = UseCaseExecutor.getInstance();
        saveHourlyDates = new SaveHourlyTiming();
        saveDailyRental = new SaveDailyRental();
    }

    @Override
    public void init() {

    }

    public void onBind(RentalDetails rentalDetails) {
        this.rentalDetails = rentalDetails;
        if (view != null) {
            view.setData(rentalDetails);
        }
    }

    public void updateAvailabilityTiming(String beginTime, String endTime) {
        if (rentalDetails == null) {
            return;
        }
        switch (strategy) {
            case HOURLY:
                updateHourlyTiming(beginTime, endTime);
                break;
            case DAILY:
                updateDailyTiming(beginTime, endTime);
                break;
        }
    }

    public void updateInstantBooking(boolean isInstantBooking) {
        if (rentalDetails == null) {
            return;
        }
        switch (strategy) {
            case DAILY:
                updateDailyInstantBooking(isInstantBooking);
                break;
            case HOURLY:
                updateHourlyInstantBooking(isInstantBooking);
                break;
        }
    }

    public void updateInstantBookingCount(int count) {
        if (rentalDetails == null) {
            return;
        }
        switch (strategy) {
            case DAILY:
                updateDailyInstantBookingCount(count);
                break;
            case HOURLY:
                updateHourlyInstantBookingCount(count);
                break;
        }
    }

    public void updateCurbsideDelivery(boolean isCurbsideDelivery) {
        if (rentalDetails == null) {
            return;
        }
        switch (strategy) {
            case DAILY:
                updateDailyCurbsideDelivery(isCurbsideDelivery);
                break;
            case HOURLY:
                updateHourlyCurbsideDelivery(isCurbsideDelivery);
                break;
        }
    }

    public void updateAcceptCash(boolean acceptCash) {
        if (rentalDetails == null) {
            return;
        }
        switch (strategy) {
            case DAILY:
                updateDailyAcceptCash(acceptCash);
                break;
            case HOURLY:
                updateHourlyAcceptCash(acceptCash);
                break;
        }
    }

    private void updateDailyTiming(String pickupTiming, String returnTiming) {
        final RentalDetails requestBody = RentalDetails.from(rentalDetails);
        requestBody.setDailyTimePickup(pickupTiming);
        requestBody.setDailyTimeReturn(returnTiming);
        executor.execute(saveDailyRental, new SaveDailyRental.RequestValues(requestBody),
                new UseCase.Callback<SaveDailyRental.ResponseValues>() {
                    @Override
                    public void onSuccess(SaveDailyRental.ResponseValues response) {
                        rentalDetails = requestBody;
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    private void updateHourlyTiming(final String beginTiming, final String endTiming) {
        executor.execute(saveHourlyDates,
                new SaveHourlyTiming.RequestValues(rentalDetails.getCarId(), beginTiming, endTiming, rentalDetails.getHourlyDates()),
                new UseCase.Callback<SaveHourlyTiming.ResponseValues>() {
                    @Override
                    public void onSuccess(SaveHourlyTiming.ResponseValues response) {
                        rentalDetails.setHourlyBeginTime(beginTiming);
                        rentalDetails.setHourlyEndTime(endTiming);
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    private void updateDailyInstantBooking(boolean isInstantBooking) {
        executor.execute(new UpdateDailyInstantBookingState(),
                new UpdateDailyInstantBookingState.RequestValues(rentalDetails.getCarId(), isInstantBooking),
                new UseCase.Callback<UpdateDailyInstantBookingState.ResponseValues>() {
                    @Override
                    public void onSuccess(UpdateDailyInstantBookingState.ResponseValues response) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails.getCarId());
                        OwnerCarsRepository.getInstance().refreshCars();
                        rentalDetails.setDailyInstantBooking(isInstantBooking);
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    private void updateHourlyInstantBooking(boolean isInstantBooking) {
        executor.execute(new UpdateHourlyInstantBookingState(),
                new UpdateHourlyInstantBookingState.RequestValues(rentalDetails.getCarId(), isInstantBooking),
                new UseCase.Callback<UpdateHourlyInstantBookingState.ResponseValues>() {
                    @Override
                    public void onSuccess(UpdateHourlyInstantBookingState.ResponseValues response) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails.getCarId());
                        OwnerCarsRepository.getInstance().refreshCars();
                        rentalDetails.setHourlyInstantBooking(isInstantBooking);
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    private void updateDailyInstantBookingCount(int count) {
        executor.execute(new UpdateDailyinstantBookingCount(),
                new UpdateDailyinstantBookingCount.RequestValues(rentalDetails.getCarId(), count),
                new UseCase.Callback<UpdateDailyinstantBookingCount.ResponseValues>() {
                    @Override
                    public void onSuccess(UpdateDailyinstantBookingCount.ResponseValues response) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails.getCarId());
                        OwnerCarsRepository.getInstance().refreshCars();
                        rentalDetails.setDailyInstantBookingCount(count);
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    private void updateHourlyInstantBookingCount(int count) {
        executor.execute(new UpdateHourlyInstantBookingCount(),
                new UpdateHourlyInstantBookingCount.RequestValues(rentalDetails.getCarId(), count),
                new UseCase.Callback<UpdateHourlyInstantBookingCount.ResponseValues>() {
                    @Override
                    public void onSuccess(UpdateHourlyInstantBookingCount.ResponseValues response) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails.getCarId());
                        OwnerCarsRepository.getInstance().refreshCars();
                        rentalDetails.setHourlyInstantBookingCount(count);
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    private void updateDailyCurbsideDelivery(boolean isCurbsideDelivery) {
        executor.execute(new UpdateDailyCurbsideDeliveryState(),
                new UpdateDailyCurbsideDeliveryState.RequestValues(rentalDetails.getCarId(), isCurbsideDelivery),
                new UseCase.Callback<UpdateDailyCurbsideDeliveryState.ResponseValues>() {
                    @Override
                    public void onSuccess(UpdateDailyCurbsideDeliveryState.ResponseValues response) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails.getCarId());
                        OwnerCarsRepository.getInstance().refreshCars();
                        rentalDetails.setDailyCurbsideDelivery(isCurbsideDelivery);
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    private void updateHourlyCurbsideDelivery(boolean isCurbsideDelivery) {
        executor.execute(new UpdateHourlyCurbsideDeliveryState(),
                new UpdateHourlyCurbsideDeliveryState.RequestValues(rentalDetails.getCarId(), isCurbsideDelivery),
                new UseCase.Callback<UpdateHourlyCurbsideDeliveryState.ResponseValues>() {
                    @Override
                    public void onSuccess(UpdateHourlyCurbsideDeliveryState.ResponseValues response) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails.getCarId());
                        OwnerCarsRepository.getInstance().refreshCars();
                        rentalDetails.setHourlyCurbsideDelivery(isCurbsideDelivery);
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    private void updateDailyAcceptCash(boolean acceptCash) {
        executor.execute(new UpdateDailyAcceptCashState(),
                new UpdateDailyAcceptCashState.RequestValues(rentalDetails.getCarId(), acceptCash),
                new UseCase.Callback<UpdateDailyAcceptCashState.ResponseValues>() {
                    @Override
                    public void onSuccess(UpdateDailyAcceptCashState.ResponseValues response) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails.getCarId());
                        OwnerCarsRepository.getInstance().refreshCars();
                        rentalDetails.setDailyAcceptCash(acceptCash);
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    private void updateHourlyAcceptCash(boolean acceptCash) {
        executor.execute(new UpdateHourlyAcceptCashState(),
                new UpdateHourlyAcceptCashState.RequestValues(rentalDetails.getCarId(), acceptCash),
                new UseCase.Callback<UpdateHourlyAcceptCashState.ResponseValues>() {
                    @Override
                    public void onSuccess(UpdateHourlyAcceptCashState.ResponseValues response) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails.getCarId());
                        OwnerCarsRepository.getInstance().refreshCars();
                        rentalDetails.setHourlyAcceptCash(acceptCash);
                    }

                    @Override
                    public void onError(Error error) {
                        if (view != null) {
                            view.setData(rentalDetails);
                        }
                    }
                });
    }

    @Override
    public void onDestroy() {
        view = null;
    }
}

File: app/src/main/java/com/cardee/owner_car_details/presenter/StrategyRentalDetailPresenter.kt
package com.cardee.owner_car_details.presenter

import com.cardee.R
import com.cardee.data_source.Error
import com.cardee.data_source.OwnerCarRepository
import com.cardee.data_source.OwnerCarsRepository
import com.cardee.domain.UseCase
import com.cardee.domain.UseCaseExecutor
import com.cardee.domain.owner.entity.RentalDetails
import com.cardee.domain.owner.usecase.*
import com.cardee.owner_car_details.RentalDetailsContract

class StrategyRentalDetailPresenter(private var view: RentalDetailsContract.ControlView?, private val strategy: Strategy) : RentalDetailsContract.Presenter {
    override fun init() {
    }

    private val executor: UseCaseExecutor = UseCaseExecutor.getInstance()
    private val saveHourlyDates: SaveHourlyTiming = SaveHourlyTiming()
    private val saveDailyRental: SaveDailyRental = SaveDailyRental()
    private var rentalDetails: RentalDetails? = null

    enum class Strategy {
        DAILY, HOURLY
    }

    fun onBind(rentalDetails: RentalDetails) {
        this.rentalDetails = rentalDetails
        view?.setData(rentalDetails)
    }

    fun updateAvailabilityTiming(beginTime: String, endTime: String) {
        if (rentalDetails == null) {
            return
        }
        when (strategy) {
            StrategyRentalDetailPresenter.Strategy.HOURLY -> updateHourlyTiming(beginTime, endTime)
            StrategyRentalDetailPresenter.Strategy.DAILY -> updateDailyTiming(beginTime, endTime)
        }
    }

    fun updateInstantBooking(isInstantBooking: Boolean) {
        if (rentalDetails == null) {
            return
        }
        when (strategy) {
            StrategyRentalDetailPresenter.Strategy.DAILY -> updateDailyInstantBooking(isInstantBooking)
            StrategyRentalDetailPresenter.Strategy.HOURLY -> updateHourlyInstantBooking(isInstantBooking)
        }
    }

    fun updateInstantBookingCount(count: Int) {
        if (rentalDetails == null) {
            return
        }
        when (strategy) {
            StrategyRentalDetailPresenter.Strategy.DAILY -> updateDailyInstantBookingCount(count)
            StrategyRentalDetailPresenter.Strategy.HOURLY -> updateHourlyInstantBookingCount(count)
        }
    }

    fun updateCurbsideDelivery(isCurbsideDelivery: Boolean) {
        if (rentalDetails == null) {
            return
        }
        when (strategy) {
            StrategyRentalDetailPresenter.Strategy.DAILY -> updateDailyCurbsideDelivery(isCurbsideDelivery)
            StrategyRentalDetailPresenter.Strategy.HOURLY -> updateHourlyCurbsideDelivery(isCurbsideDelivery)
        }
    }

    fun updateMinimumBooking(minimumBookingEnabled: Boolean) {
        rentalDetails ?: return
        when (strategy) {
            StrategyRentalDetailPresenter.Strategy.DAILY -> {
                val duration =
                        if (minimumBookingEnabled) {
                            val dur = rentalDetails?.dailyMinRentalDuration
                            if (dur == null || dur < 2) {
                                2
                            } else {
                                dur
                            }
                        } else {
                            0
                        }
                updateDailyMinimumBooking(duration)
            }

            StrategyRentalDetailPresenter.Strategy.HOURLY -> {
                val duration =
                        if (minimumBookingEnabled) {
                            val dur = rentalDetails?.hourlyMinRentalDuration
                            if (dur == null || dur < 2) {
                                2
                            } else {
                                dur
                            }
                        } else {
                            0
                        }
                updateHourlyMinimumBooking(duration)
            }
        }
    }

    fun updateDailyMinimumBooking(minimumDuration: Int) {
        executor.execute(UpdateDailyMinRentDuration(),
                UpdateDailyMinRentDuration.RequestValues(OwnerCarRepository.currentCarId(), minimumDuration),
                object : UseCase.Callback<UpdateDailyMinRentDuration.ResponseValues> {
                    override fun onSuccess(response: UpdateDailyMinRentDuration.ResponseValues?) {
                        view?.showProgress(false)
                        if (response?.isSuccess == true) {
                            OwnerCarRepository.getInstance().refresh(OwnerCarRepository.currentCarId())
                            OwnerCarsRepository.getInstance().refreshCars()
                            view?.showMessage(R.string.saved_successfully)
                            rentalDetails?.dailyMinRentalDuration = minimumDuration
                            view?.setMinBookingValue(minimumDuration)
                        }
                    }

                    override fun onError(error: Error?) {
                        view?.showProgress(false)
                        view?.showMessage(error?.message)
                    }
                })
    }

    fun updateHourlyMinimumBooking(minimumDuration: Int) {
        executor.execute(UpdateHourlyMinRentDuration(),
                UpdateHourlyMinRentDuration.RequestValues(OwnerCarRepository.currentCarId(), minimumDuration),
                object : UseCase.Callback<UpdateHourlyMinRentDuration.ResponseValues> {
                    override fun onSuccess(response: UpdateHourlyMinRentDuration.ResponseValues?) {
                        view?.showProgress(false)
                        if (response?.isSuccess == true) {
                            OwnerCarRepository.getInstance().refresh(OwnerCarRepository.currentCarId())
                            OwnerCarsRepository.getInstance().refreshCars()
                            view?.showMessage(R.string.saved_successfully)
                            rentalDetails?.hourlyMinRentalDuration = minimumDuration
                            view?.setMinBookingValue(minimumDuration)
                        }
                    }

                    override fun onError(error: Error?) {
                        view?.showProgress(false)
                        view?.showMessage(error?.message)
                    }
                })
    }

    fun updateAcceptCash(acceptCash: Boolean) {
        if (rentalDetails == null) {
            return
        }
        when (strategy) {
            StrategyRentalDetailPresenter.Strategy.DAILY -> updateDailyAcceptCash(acceptCash)
            StrategyRentalDetailPresenter.Strategy.HOURLY -> updateHourlyAcceptCash(acceptCash)
        }
    }

    private fun updateDailyTiming(pickupTiming: String, returnTiming: String) {
        val requestBody = RentalDetails.from(rentalDetails!!)
        requestBody.dailyTimePickup = pickupTiming
        requestBody.dailyTimeReturn = returnTiming
        executor.execute<SaveDailyRental.RequestValues, SaveDailyRental.ResponseValues>(saveDailyRental, SaveDailyRental.RequestValues(requestBody),
                object : UseCase.Callback<SaveDailyRental.ResponseValues> {
                    override fun onSuccess(response: SaveDailyRental.ResponseValues) {
                        rentalDetails = requestBody
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    private fun updateHourlyTiming(beginTiming: String, endTiming: String) {
        executor.execute<SaveHourlyTiming.RequestValues, SaveHourlyTiming.ResponseValues>(saveHourlyDates,
                SaveHourlyTiming.RequestValues(rentalDetails!!.carId, beginTiming, endTiming, rentalDetails!!.hourlyDates),
                object : UseCase.Callback<SaveHourlyTiming.ResponseValues> {
                    override fun onSuccess(response: SaveHourlyTiming.ResponseValues) {
                        rentalDetails!!.hourlyBeginTime = beginTiming
                        rentalDetails!!.hourlyEndTime = endTiming
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    private fun updateDailyInstantBooking(isInstantBooking: Boolean) {
        executor.execute<UpdateDailyInstantBookingState.RequestValues, UpdateDailyInstantBookingState.ResponseValues>(UpdateDailyInstantBookingState(),
                UpdateDailyInstantBookingState.RequestValues(rentalDetails!!.carId, isInstantBooking),
                object : UseCase.Callback<UpdateDailyInstantBookingState.ResponseValues> {
                    override fun onSuccess(response: UpdateDailyInstantBookingState.ResponseValues) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails!!.carId)
                        OwnerCarsRepository.getInstance().refreshCars()
                        rentalDetails!!.isDailyInstantBooking = isInstantBooking
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    private fun updateHourlyInstantBooking(isInstantBooking: Boolean) {
        executor.execute<UpdateHourlyInstantBookingState.RequestValues, UpdateHourlyInstantBookingState.ResponseValues>(UpdateHourlyInstantBookingState(),
                UpdateHourlyInstantBookingState.RequestValues(rentalDetails!!.carId, isInstantBooking),
                object : UseCase.Callback<UpdateHourlyInstantBookingState.ResponseValues> {
                    override fun onSuccess(response: UpdateHourlyInstantBookingState.ResponseValues) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails!!.carId)
                        OwnerCarsRepository.getInstance().refreshCars()
                        rentalDetails!!.isHourlyInstantBooking = isInstantBooking
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    private fun updateDailyInstantBookingCount(count: Int) {
        executor.execute<UpdateDailyinstantBookingCount.RequestValues, UpdateDailyinstantBookingCount.ResponseValues>(UpdateDailyinstantBookingCount(),
                UpdateDailyinstantBookingCount.RequestValues(rentalDetails!!.carId, count),
                object : UseCase.Callback<UpdateDailyinstantBookingCount.ResponseValues> {
                    override fun onSuccess(response: UpdateDailyinstantBookingCount.ResponseValues) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails!!.carId)
                        OwnerCarsRepository.getInstance().refreshCars()
                        rentalDetails!!.dailyInstantBookingCount = count
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    private fun updateHourlyInstantBookingCount(count: Int) {
        executor.execute<UpdateHourlyInstantBookingCount.RequestValues, UpdateHourlyInstantBookingCount.ResponseValues>(UpdateHourlyInstantBookingCount(),
                UpdateHourlyInstantBookingCount.RequestValues(rentalDetails!!.carId, count),
                object : UseCase.Callback<UpdateHourlyInstantBookingCount.ResponseValues> {
                    override fun onSuccess(response: UpdateHourlyInstantBookingCount.ResponseValues) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails!!.carId)
                        OwnerCarsRepository.getInstance().refreshCars()
                        rentalDetails!!.hourlyInstantBookingCount = count
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    private fun updateDailyCurbsideDelivery(isCurbsideDelivery: Boolean) {
        executor.execute<UpdateDailyCurbsideDeliveryState.RequestValues, UpdateDailyCurbsideDeliveryState.ResponseValues>(UpdateDailyCurbsideDeliveryState(),
                UpdateDailyCurbsideDeliveryState.RequestValues(rentalDetails!!.carId, isCurbsideDelivery),
                object : UseCase.Callback<UpdateDailyCurbsideDeliveryState.ResponseValues> {
                    override fun onSuccess(response: UpdateDailyCurbsideDeliveryState.ResponseValues) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails!!.carId)
                        OwnerCarsRepository.getInstance().refreshCars()
                        rentalDetails!!.isDailyCurbsideDelivery = isCurbsideDelivery
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    private fun updateHourlyCurbsideDelivery(isCurbsideDelivery: Boolean) {
        executor.execute<UpdateHourlyCurbsideDeliveryState.RequestValues, UpdateHourlyCurbsideDeliveryState.ResponseValues>(UpdateHourlyCurbsideDeliveryState(),
                UpdateHourlyCurbsideDeliveryState.RequestValues(rentalDetails!!.carId, isCurbsideDelivery),
                object : UseCase.Callback<UpdateHourlyCurbsideDeliveryState.ResponseValues> {
                    override fun onSuccess(response: UpdateHourlyCurbsideDeliveryState.ResponseValues) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails!!.carId)
                        OwnerCarsRepository.getInstance().refreshCars()
                        rentalDetails!!.isHourlyCurbsideDelivery = isCurbsideDelivery
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    private fun updateDailyAcceptCash(acceptCash: Boolean) {
        executor.execute<UpdateDailyAcceptCashState.RequestValues, UpdateDailyAcceptCashState.ResponseValues>(UpdateDailyAcceptCashState(),
                UpdateDailyAcceptCashState.RequestValues(rentalDetails!!.carId, acceptCash),
                object : UseCase.Callback<UpdateDailyAcceptCashState.ResponseValues> {
                    override fun onSuccess(response: UpdateDailyAcceptCashState.ResponseValues) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails!!.carId)
                        OwnerCarsRepository.getInstance().refreshCars()
                        rentalDetails!!.isDailyAcceptCash = acceptCash
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    private fun updateHourlyAcceptCash(acceptCash: Boolean) {
        executor.execute<UpdateHourlyAcceptCashState.RequestValues, UpdateHourlyAcceptCashState.ResponseValues>(UpdateHourlyAcceptCashState(),
                UpdateHourlyAcceptCashState.RequestValues(rentalDetails!!.carId, acceptCash),
                object : UseCase.Callback<UpdateHourlyAcceptCashState.ResponseValues> {
                    override fun onSuccess(response: UpdateHourlyAcceptCashState.ResponseValues) {
                        OwnerCarRepository.getInstance().refresh(rentalDetails!!.carId)
                        OwnerCarsRepository.getInstance().refreshCars()
                        rentalDetails!!.isHourlyAcceptCash = acceptCash
                    }

                    override fun onError(error: Error) {
                        if (view != null) {
                            view!!.setData(rentalDetails)
                        }
                    }
                })
    }

    override fun onDestroy() {
        view = null
    }
}
